File: prototype/sarathi/runtime/scheduler.py
"""Deciding when NOT to run the model.

Everything else in this project makes inference better. This module makes it
happen less often, which is the only reason the app can run for a day.

The ordering of the gates is the design. They run cheapest-first, so a frame
that will be skipped is skipped for the least possible cost:

1. **Staleness** - free. A frame that queued behind a slow inference describes
   a scene the user has walked past.
2. **Motion gate** - about 0.1 ms. A stationary user pointed at a static scene
   needs no inference at all, and indoors that is most frames.
3. **Rate limit** - free. Even when things are changing, there is no value in
   running faster than the user can act on.
4. **Thermal** - cached. Degrade before the OS throttles, not after.

The gate is run at a *higher* rate than inference on purpose. Frame
differencing on a 64x64 downscale costs microseconds; the detector costs tens
of milliseconds. Sampling motion cheaply and often is what lets the inference
rate drop to idle without the system being slow to notice the user has started
walking again.

One safeguard that matters more than it looks: a keepalive. Pure motion gating
can wedge - a miscalibrated threshold, a camera that returns a frozen frame, a
scene that changes in ways the downscale cannot see - and a blind user would
have no way to tell the difference between "nothing to report" and "stopped
working". So inference runs at a floor rate regardless of what the gates say.
"""

from __future__ import annotations

import shutil
import subprocess
import time
from dataclasses import dataclass, field
from enum import Enum

import cv2
import numpy as np

from ..types import Frame
from ..util.log import get_logger

log = get_logger(__name__)


class Activity(Enum):
    """What the user appears to be doing, inferred from the scene."""

    IDLE = "idle"
    MOVING = "moving"


class Skip(Enum):
    """Why a frame was not processed. Recorded so the docs can show real
    proportions rather than claimed ones."""

    RAN = "ran"
    STALE = "stale"
    STATIC = "static"
    RATE = "rate_limited"
    THERMAL = "thermal"


@dataclass
class Decision:
    run: bool
    reason: Skip
    activity: Activity
    target_hz: float
    #: True when this pass is a keepalive rather than a motion-triggered one.
    keepalive: bool = False


# -- motion ------------------------------------------------------------------


class MotionGate:
    """Frame-difference gate on a small greyscale downscale.

    Deliberately crude. The question is not "what moved" but "is this worth
    looking at properly", and a 64x64 mean absolute difference answers that for
    a few microseconds. Anything more sophisticated would cost a meaningful
    fraction of the inference it is meant to avoid.
    """

    def __init__(self, *, threshold: float = 0.012, size: int = 64) -> None:
        self.threshold = threshold
        self.size = size
        self._previous: np.ndarray | None = None
        self.last_score: float = 1.0

    def score(self, image: np.ndarray) -> float:
        """Normalised mean absolute difference from the previous frame."""
        small = cv2.resize(image, (self.size, self.size), interpolation=cv2.INTER_AREA)
        if small.ndim == 3:
            small = cv2.cvtColor(small, cv2.COLOR_BGR2GRAY)
        current = small.astype(np.float32) / 255.0

        if self._previous is None:
            self._previous = current
            # The first frame has nothing to compare against. Treat it as
            # changed so the pipeline always produces one result at startup
            # rather than sitting silent until something moves.
            self.last_score = 1.0
            return self.last_score

        self.last_score = float(np.abs(current - self._previous).mean())
        self._previous = current
        return self.last_score

    def changed(self, image: np.ndarray) -> bool:
        return self.score(image) >= self.threshold

    def reset(self) -> None:
        self._previous = None
        self.last_score = 1.0


# -- thermal -----------------------------------------------------------------


class ThermalReader:
    """Headroom before the platform starts throttling us.

    Returns 0.0 (cool) to 1.0 (about to be throttled). Android has
    PowerManager.getThermalHeadroom for exactly this; the desktop
    implementations here exist so the policy can be developed and tested
    before the port.
    """

    def pressure(self) -> float:
        return 0.0


class NullThermalReader(ThermalReader):
    """Always cool. Default under test and on platforms with no signal."""


class MacThermalReader(ThermalReader):
    """Reads macOS CPU speed limit via `pmset -g therm`.

    Not equivalent to Android's headroom - it reports throttling that is
    already happening rather than how close we are to it - so it is a
    development stand-in, not the real policy input. The Android reader is what
    the shipped behaviour will be tuned against.
    """

    def __init__(self, cache_s: float = 2.0) -> None:
        self.cache_s = cache_s
        self._value = 0.0
        self._checked_at = -1e9
        self._available = shutil.which("pmset") is not None

    def pressure(self) -> float:
        if not self._available:
            return 0.0
        now = time.monotonic()
        if now - self._checked_at < self.cache_s:
            return self._value
        self._checked_at = now
        try:
            out = subprocess.run(
                ["pmset", "-g", "therm"], capture_output=True, text=True, timeout=2
            ).stdout
        except (OSError, subprocess.SubprocessError):
            self._available = False
            return 0.0
        for line in out.splitlines():
            if "CPU_Speed_Limit" in line:
                _, _, value = line.partition("=")
                try:
                    limit = float(value.strip())
                except ValueError:
                    break
                self._value = max(0.0, min(1.0, (100.0 - limit) / 100.0))
                return self._value
        self._value = 0.0
        return self._value


# -- scheduler ---------------------------------------------------------------


@dataclass
class SchedulerConfig:
    #: Inference rate while the user is moving.
    max_inference_hz: float = 8.0
    #: Inference rate while stationary and the scene is static.
    idle_inference_hz: float = 1.0
    #: Floor rate that runs regardless of gating. The safeguard against a
    #: wedged gate leaving the user silently unguided.
    keepalive_hz: float = 0.2
    #: Frames older than this at decision time are discarded.
    max_frame_age_ms: float = 250.0

    motion_enabled: bool = True
    motion_threshold: float = 0.012
    motion_size: int = 64
    #: How long the scene must stay static before dropping to the idle rate.
    settle_s: float = 2.0

    thermal_enabled: bool = True
    #: Above this pressure, start shedding frame rate.
    thermal_soft: float = 0.30
    #: Above this, drop to the idle rate whatever the user is doing.
    thermal_hard: float = 0.70

    depth_hz: float = 2.0
    #: Depth only earns its cost while the user is actually moving.
    depth_when_idle: bool = False


@dataclass
class SchedulerStats:
    considered: int = 0
    ran: int = 0
    keepalives: int = 0
    skipped: dict[str, int] = field(default_factory=dict)
    #: Rolling mean of measured inference cost, used to report duty cycle.
    inference_ms: float = 0.0
    _inference_samples: int = 0

    def note_skip(self, reason: Skip) -> None:
        self.skipped[reason.value] = self.skipped.get(reason.value, 0) + 1

    def note_inference(self, ms: float) -> None:
        self._inference_samples += 1
        n = self._inference_samples
        self.inference_ms += (ms - self.inference_ms) / n

    @property
    def skip_rate(self) -> float:
        return 0.0 if self.considered == 0 else 1.0 - self.ran / self.considered

    def summary(self) -> str:
        lines = [
            f"frames considered {self.considered}",
            f"inference ran     {self.ran}  ({100 * (1 - self.skip_rate):.1f}%)",
            f"  of which keepalive {self.keepalives}",
            f"mean inference    {self.inference_ms:.1f} ms",
        ]
        for reason, count in sorted(self.skipped.items(), key=lambda kv: -kv[1]):
            share = 100 * count / max(1, self.considered)
            lines.append(f"skipped {reason:<13} {count:>6}  ({share:.1f}%)")
        return "\n".join(lines)


class Scheduler:
    """Decides, per frame, whether to spend inference on it."""

    #: Accept frames arriving up to 5% early against the rate target. See the
    #: rate-limit branch in `decide` for why a strict comparison is wrong.
    _RATE_TOLERANCE = 0.95

    def __init__(
        self,
        config: SchedulerConfig | None = None,
        thermal: ThermalReader | None = None,
    ) -> None:
        self.config = config or SchedulerConfig()
        self.thermal = thermal or NullThermalReader()
        self.gate = MotionGate(
            threshold=self.config.motion_threshold, size=self.config.motion_size
        )
        self.stats = SchedulerStats()

        self._last_inference = -1e9
        self._last_depth = -1e9
        self._last_motion_at = -1e9
        self._activity = Activity.MOVING  # assume moving until proven otherwise

    @property
    def activity(self) -> Activity:
        return self._activity

    def target_hz(self, pressure: float) -> float:
        cfg = self.config
        base = cfg.max_inference_hz if self._activity is Activity.MOVING else cfg.idle_inference_hz
        if not cfg.thermal_enabled or pressure <= cfg.thermal_soft:
            return base
        if pressure >= cfg.thermal_hard:
            return cfg.idle_inference_hz
        # Linear shed between soft and hard, so the rate falls off gradually
        # instead of stepping down in a way the user would hear.
        span = max(1e-6, cfg.thermal_hard - cfg.thermal_soft)
        fraction = (pressure - cfg.thermal_soft) / span
        return base - (base - cfg.idle_inference_hz) * fraction

    def decide(self, frame: Frame, now: float | None = None) -> Decision:
        cfg = self.config
        now = time.monotonic() if now is None else now
        self.stats.considered += 1

        # 1. Staleness. Free, and acting on a stale frame is worse than not
        #    acting: it describes somewhere the user no longer is.
        if frame.age_ms(now) > cfg.max_frame_age_ms:
            self.stats.note_skip(Skip.STALE)
            return Decision(False, Skip.STALE, self._activity, 0.0)

        # 2. Motion. Cheap enough to run on nearly every frame, which is what
        #    lets the inference rate idle down without being slow to wake.
        moved = True
        if cfg.motion_enabled:
            moved = self.gate.changed(frame.image)
            if moved:
                self._last_motion_at = now
            self._activity = (
                Activity.MOVING if now - self._last_motion_at < cfg.settle_s else Activity.IDLE
            )

        pressure = self.thermal.pressure() if cfg.thermal_enabled else 0.0
        hz = self.target_hz(pressure)
        since = now - self._last_inference

        # 3. Keepalive. Checked before the gates so nothing can suppress it.
        if cfg.keepalive_hz > 0 and since >= 1.0 / cfg.keepalive_hz:
            self._last_inference = now
            self.stats.ran += 1
            self.stats.keepalives += 1
            return Decision(True, Skip.RAN, self._activity, hz, keepalive=True)

        # 4. Rate limit, with a small tolerance for early arrival.
        #
        #    Frames never land on exact intervals. Compared strictly, a camera
        #    running at twice the target rate drops to two-thirds of it: the
        #    frame at +0.199999 s is rejected, so the next candidate is at
        #    +0.3 s, and the pattern alternates. The tolerance costs nothing
        #    and recovers the rate that was actually asked for.
        if hz <= 0 or since < (1.0 / hz) * self._RATE_TOLERANCE:
            reason = (
                Skip.THERMAL
                if cfg.thermal_enabled and pressure > cfg.thermal_soft
                else Skip.RATE
            )
            self.stats.note_skip(reason)
            return Decision(False, reason, self._activity, hz)

        # 5. Static scene. Last because it is the most expensive check, and
        #    because a frame that fails the rate limit never needed it.
        if cfg.motion_enabled and not moved and self._activity is Activity.IDLE:
            self.stats.note_skip(Skip.STATIC)
            return Decision(False, Skip.STATIC, self._activity, hz)

        self._last_inference = now
        self.stats.ran += 1
        return Decision(True, Skip.RAN, self._activity, hz)

    def should_run_depth(self, now: float | None = None) -> bool:
        """Whether the Tier 2 depth pass has earned its cost this frame."""
        cfg = self.config
        now = time.monotonic() if now is None else now
        if cfg.depth_hz <= 0:
            return False
        if not cfg.depth_when_idle and self._activity is Activity.IDLE:
            return False
        if now - self._last_depth < 1.0 / cfg.depth_hz:
            return False
        self._last_depth = now
        return True

    def note_inference(self, duration_ms: float) -> None:
        self.stats.note_inference(duration_ms)

    def reset(self) -> None:
        self.gate.reset()
        self.stats = SchedulerStats()
        self._last_inference = -1e9
        self._last_depth = -1e9
        self._last_motion_at = -1e9
        self._activity = Activity.MOVING

File: android/app/src/main/java/in/sarathi/app/runtime/Scheduler.kt
package `in`.sarathi.app.runtime

import android.os.Build
import android.os.PowerManager
import kotlin.math.abs

/**
 * Deciding when NOT to run the model.
 *
 * Everything else makes inference better; this makes it happen less often,
 * which is the only reason the app can run for a day. Measured on the desktop
 * prototype: a stationary user needs inference on 3% of frames, a walking one
 * on 18%.
 *
 * Gates run cheapest-first so a frame that will be skipped is skipped for the
 * least possible cost: staleness, then motion at about 0.1 ms, then rate, then
 * thermal.
 */
class Scheduler(
    private val config: Config = Config(),
    private val power: PowerManager? = null,
) {
    data class Config(
        val maxInferenceHz: Double = 8.0,
        val idleInferenceHz: Double = 1.0,
        /**
         * Floor rate that runs regardless of any gate. A blind user cannot
         * distinguish "nothing to report" from "stopped working", so no gate
         * is permitted to silence the system completely.
         */
        val keepaliveHz: Double = 0.2,
        val maxFrameAgeMs: Long = 250,
        val motionEnabled: Boolean = true,
        val motionThreshold: Double = 0.012,
        val settleMs: Long = 2000,
        val thermalEnabled: Boolean = true,
        val thermalSoft: Double = 0.30,
        val thermalHard: Double = 0.70,
        val depthHz: Double = 2.0,
    )

    enum class Skip { RAN, STALE, STATIC, RATE, THERMAL }
    enum class Activity { IDLE, MOVING }

    data class Decision(
        val run: Boolean,
        val reason: Skip,
        val activity: Activity,
        val targetHz: Double,
        val keepalive: Boolean = false,
    )

    private val gate = MotionGate(config.motionThreshold)
    private var lastInference = 0L
    private var lastDepth = 0L
    private var lastMotionAt = 0L
    var activity: Activity = Activity.MOVING
        private set

    var framesConsidered = 0L; private set
    var framesRan = 0L; private set
    val skips = mutableMapOf<Skip, Long>()

    /**
     * Thermal headroom, 0 cool to 1 about to throttle.
     *
     * Android exposes this directly from API 30. Degrading before the platform
     * throttles keeps the announcement rate falling smoothly instead of the OS
     * yanking it - a step change in how often the system speaks is audible and
     * unsettling; a ramp is not.
     */
    private fun thermalPressure(): Double {
        val pm = power
        if (!config.thermalEnabled || pm == null) return 0.0
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return 0.0
        val headroom = runCatching { pm.getThermalHeadroom(10) }.getOrNull() ?: return 0.0
        if (headroom.isNaN()) return 0.0
        return headroom.toDouble().coerceIn(0.0, 1.0)
    }

    fun targetHz(pressure: Double): Double {
        val base = if (activity == Activity.MOVING) config.maxInferenceHz else config.idleInferenceHz
        if (!config.thermalEnabled || pressure <= config.thermalSoft) return base
        if (pressure >= config.thermalHard) return config.idleInferenceHz
        val span = (config.thermalHard - config.thermalSoft).coerceAtLeast(1e-6)
        val fraction = (pressure - config.thermalSoft) / span
        return base - (base - config.idleInferenceHz) * fraction
    }

    /**
     * @param luma downscaled greyscale plane, for the motion gate
     * @param frameAgeMs how stale this frame already is
     */
    fun decide(luma: ByteArray, width: Int, height: Int, frameAgeMs: Long, nowMs: Long): Decision {
        framesConsidered++

        if (frameAgeMs > config.maxFrameAgeMs) return skip(Skip.STALE, 0.0)

        var moved = true
        if (config.motionEnabled) {
            moved = gate.changed(luma, width, height)
            if (moved) lastMotionAt = nowMs
            activity = if (nowMs - lastMotionAt < config.settleMs) Activity.MOVING else Activity.IDLE
        }

        val pressure = thermalPressure()
        val hz = targetHz(pressure)
        val since = nowMs - lastInference

        // Keepalive is checked before the gates so nothing can suppress it.
        if (config.keepaliveHz > 0 && since >= (1000.0 / config.keepaliveHz)) {
            lastInference = nowMs
            framesRan++
            return Decision(true, Skip.RAN, activity, hz, keepalive = true)
        }

        // 5% tolerance: frames never land on exact intervals, and comparing
        // strictly costs about a fifth of the target rate - a camera at twice
        // the target ends up delivering two thirds of it.
        if (hz <= 0 || since < (1000.0 / hz) * 0.95) {
            val reason = if (config.thermalEnabled && pressure > config.thermalSoft)
                Skip.THERMAL else Skip.RATE
            return skip(reason, hz)
        }

        if (config.motionEnabled && !moved && activity == Activity.IDLE) return skip(Skip.STATIC, hz)

        lastInference = nowMs
        framesRan++
        return Decision(true, Skip.RAN, activity, hz)
    }

    /** Whether the Tier 2 depth pass has earned its cost. */
    fun shouldRunDepth(nowMs: Long): Boolean {
        if (config.depthHz <= 0) return false
        if (activity == Activity.IDLE) return false   // steps do not appear while standing still
        if (nowMs - lastDepth < 1000.0 / config.depthHz) return false
        lastDepth = nowMs
        return true
    }

    private fun skip(reason: Skip, hz: Double): Decision {
        skips[reason] = (skips[reason] ?: 0) + 1
        return Decision(false, reason, activity, hz)
    }

    val skipRate: Double
        get() = if (framesConsidered == 0L) 0.0 else 1.0 - framesRan.toDouble() / framesConsidered
}

/**
 * Frame-difference gate on a small greyscale downscale.
 *
 * Deliberately crude: the question is not "what moved" but "is this worth
 * looking at properly", and anything more sophisticated would cost a
 * meaningful fraction of the inference it exists to avoid.
 */
class MotionGate(private val threshold: Double, private val size: Int = 64) {
    private var previous: FloatArray? = null
    var lastScore: Double = 1.0
        private set

    fun changed(luma: ByteArray, width: Int, height: Int): Boolean {
        val current = FloatArray(size * size)
        val sx = width.toFloat() / size
        val sy = height.toFloat() / size
        for (y in 0 until size) {
            val srcY = (y * sy).toInt().coerceIn(0, height - 1)
            for (x in 0 until size) {
                val srcX = (x * sx).toInt().coerceIn(0, width - 1)
                val index = srcY * width + srcX
                current[y * size + x] =
                    (luma.getOrElse(index) { 0 }.toInt() and 0xFF) / 255f
            }
        }
        val prev = previous
        previous = current
        if (prev == null) {
            // Nothing to compare against. Treat the first frame as changed so
            // the pipeline always produces one result at startup rather than
            // sitting silent until something moves.
            lastScore = 1.0
            return true
        }
        var sum = 0.0
        for (i in current.indices) sum += abs(current[i] - prev[i])
        lastScore = sum / current.size
        return lastScore >= threshold
    }

    fun reset() { previous = null; lastScore = 1.0 }
}
